### Grid Search
#### Machine Learning Algorithms has parameters that control the complexity how models learn. For Example logistic regression from sklearn module has parameter C which control complexity of the model. Here we learn that using grid search that provide us results on different C values.

### Grid Search
##### So we apply grid search which provide us to insert different values on the result basis we chose our best value. Grid search allow us to try different values. We can evaluate all the combination of sets. This forming a grid of values.
##### Higher values of C tell the model, the training data resembles real world information, place a greater weight on the training data. While lower values of C do the opposite.

In [217]:
import numpy as np
import pandas as pd
from sklearn import datasets
iris = datasets.load_iris()

In [218]:
# let's see the type
type(iris)

sklearn.utils._bunch.Bunch

In [219]:
# # if you'd like to view list of attributes use
dir(dataset_iris)

['DESCR',
 'data',
 'data_module',
 'feature_names',
 'filename',
 'frame',
 'target',
 'target_names']

In [220]:
#### Changing Sklearn into pandas dataset

In [221]:
# np.c_ is the numpy concatenate function
# which is used to concat iris['data'] and iris['target'] arrays
# for pandas column argument: concat iris['feature_names'] list
# and string list (in this case one string); you can make this anything you'd like..
# the original dataset would probably call this ['Species']
data1 = pd.DataFrame(data= np.c_[iris['data'], iris['target']],
                     columns= iris['feature_names'] + ['target'])

In [222]:
data1.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0     0.0  
1     0.0  
2     0.0  
3     0.0  
4     0.0

In [223]:
data1.shape

(150, 5)

In [224]:
data1.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

#### Second Method

In [225]:
from sklearn.datasets import load_iris
import pandas as pd

data_2_method = load_iris()
df = pd.DataFrame(data=data_2_method.data, columns=data_2_method.feature_names)
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

#### lets tele without pandas and complete the grid search learning

In [226]:
# Next in order to create the model we must have a set of independent variables X and a dependant variable y.
X = iris['data']
y = iris['target']

#### let's use model
#### Now we will load the logistic model for classifying the iris flowers.

In [227]:
from sklearn.linear_model import LogisticRegression

#### Creating the model, setting max_iter to a higher value to ensure that the model finds a result.

#### Keep in mind the default value for C in a logistic regression model is 1, we will compare this later.

#### In the example below, we look at the iris data set and try to train a model with varying values for C in logistic regression.

In [228]:
logit = LogisticRegression(max_iter = 10000)

In [229]:
# After we create the model, we must fit the model to the data.
print(logit.fit(X,y))

LogisticRegression(max_iter=10000)


In [230]:
# To evaluate the model we run the score method.
print(logit.score(X,y))

0.9733333333333334


- With the default setting of C = 1, we achieved a score of 0.973.
- Let's see if we can do any better by implementing a grid search with difference values of 0.973.


## Now let's implement Grid Search.

In [231]:
# first we set ranges of C
# Since the default value for C is 1, we will set a range of values surrounding it.
C = [0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2]

In [232]:
# Now we create a loop that will change value of C and evaluate the model with each change. First we create an empty list to store the score within
scores = [] # this will store the values for different C value.

In [233]:
# To change the values we will loop over the range of values and then update the parameter each time
# let's see how we can do it.
for choice in C:
    logit.set_params(C=choice)
    logit.fit(X,y)
    scores.append(logit.score(X,y))
print(scores)

[0.9666666666666667, 0.9666666666666667, 0.9733333333333334, 0.9733333333333334, 0.98, 0.98, 0.9866666666666667, 0.9866666666666667]


- Now you can see different values when we change the value of C
- Increasing the value of C give the best result. Where best results is 0.9866666666666667 when C is 1.75

* We can see that the lower values of C performed worse than the base parameter of 1. However, as we increased the value of C to 1.75 the model experienced increased accuracy.

* It seems that increasing C beyond this amount does not help increase model accuracy.

## Summary
##### To evaluate the performance of our logistic regression model, we made use of the same data that was employed during its training. If the model is too accurate in its representation of those facts, then its ability to forecast previously unknown data may be limited. The term for this type of statistical mistake is "over fitting."

##### We can avoid being mislead by the scores on the training data by setting aside a piece of our data and using it especially for the purpose of testing the model. We can do this by setting aside a portion of our data and using it exclusively for the purpose of testing the model. If you want to avoid being mislead and overfitting,

#### Future Work
##### Next time I will try fit new model and see evaluate the result of C and also try train_test split to see how our model work when unseen data is also applied